src/algs/direct/DIRect.c

```diff
@@ -80,40 +78,91 @@
     doublereal *levels = 0, *thirds = 0;
     doublereal epsfix;
     integer oldpos, minpos, maxpos, tstart, actdeep, ifreeold, oldmaxf;
-    integer numfunc, version;
+    integer version;
     integer jones;
 
-    /* FIXME: change sizes dynamically? */
-#define MY_ALLOC(p, t, n) p = (t *) malloc(sizeof(t) * (n)); \
-                          if (!(p)) { *ierror = -100; goto cleanup; }
-
-    /* Note that I've transposed c__, length, and f relative to the 
-       original Fortran code.  e.g. length was length(maxfunc,n) 
+    /* Note that I've transposed c__, length, and f relative to the
+       original Fortran code.  e.g. length was length(maxfunc,n)
        in Fortran [ or actually length(maxfunc, maxdims), but by
        using malloc I can just allocate n ], corresponding to
        length[n][maxfunc] in C, but I've changed the code to access
        it as length[maxfunc][n].  That is, the maxfunc direction
        is the discontiguous one.  This makes it easier to resize
        dynamically (by adding contiguous rows) using realloc, without
        having to move data around manually. */
-    MY_ALLOC(c__, doublereal, MAXFUNC * (*n));
-    MY_ALLOC(length, integer, MAXFUNC * (*n));
-    MY_ALLOC(f, doublereal, MAXFUNC * 2);
-    MY_ALLOC(point, integer, MAXFUNC);
-    if (*maxf <= 0) *maxf = MAXFUNC - 1000;
+    #define MAXMEMORY 1073741824
+    integer MAXFUNC = *maxf <= 0 ? 101000 : (*maxf + 1000 + *maxf / 2);
+    integer fixed_memory_dim = ((*n) * (sizeof(doublereal) + sizeof(integer)) +
+                                      (sizeof(doublereal) * 2 + sizeof(integer)));
+    MAXFUNC = MAXFUNC * fixed_memory_dim > MAXMEMORY ? MAXMEMORY/fixed_memory_dim : MAXFUNC;
+    c__ = (doublereal *) malloc(sizeof(doublereal) * (MAXFUNC * (*n)));
+    if (!(c__)) {
+        *ierror = -100;
+        goto cleanup;
+    }
```

Comment:
This part was added as the algorithm uses absurd amounts of memory for large number of function evaluations. The `MY_ALLOC` macro did not help, we got a segmentation fault. Could this be of interest to nlopt @stevengj ?